src/PowerShellEditorServices/Workspace/WorkspaceFileSystemWrapper.cs

```diff
@@ -0,0 +1,328 @@
+            }
+            catch (Exception e)
+            {
+                this.logger.WriteHandledException(
+                    $"Could not enumerate files in the path '{dirInfo.FullName}' due to an exception",
+                    e);
+                yield break;
+            }
+            if (result == null) { yield break; }
+            foreach (var item in result)
+            {
+                if (this.ignoreReparsePoints && item.Attributes.HasFlag(FileAttributes.ReparsePoint)) { continue; }
+                // If it's a directory then automatically return it
+                if (item.Attributes.HasFlag(FileAttributes.Directory)) { yield return item; continue; }
+                if (this.allowedExtensions == null || this.allowedExtensions.Length == 0) { yield return item; continue; }
+                foreach (string extension in allowedExtensions)
+                {
+                    if (item.Extension == extension) { yield return item; break; }
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Wraps an instance of <see cref="System.IO.DirectoryInfo" /> and provides implementation of
+    /// <see cref="Microsoft.Extensions.FileSystemGlobbing.Abstractions.DirectoryInfoBase" />.
+    /// Based on https://github.com/aspnet/Extensions/blob/c087cadf1dfdbd2b8785ef764e5ef58a1a7e5ed0/src/FileSystemGlobbing/src/Abstractions/DirectoryInfoWrapper.cs
+    /// </summary>
+    public class WorkspaceFileSystemDirectoryWrapper : DirectoryInfoBase
+    {
+        private readonly DirectoryInfo concreteDirectoryInfo;
+        private readonly bool isParentPath;
+        private readonly WorkspaceFileSystemWrapperFactory factory;
+        private readonly int depth;
+
+        /// <summary>
+        /// Initializes an instance of <see cref="WorkspaceFileSystemDirectoryWrapper" />.
+        /// </summary>
+        public WorkspaceFileSystemDirectoryWrapper(WorkspaceFileSystemWrapperFactory factory, DirectoryInfo directoryInfo, int depth)
+        {
+            this.concreteDirectoryInfo = directoryInfo;
+            this.isParentPath = (depth == 0);
+            this.factory = factory;
+            this.depth = depth;
+        }
+
+        /// <inheritdoc />
+        public override IEnumerable<FileSystemInfoBase> EnumerateFileSystemInfos()
+        {
+            if (!concreteDirectoryInfo.Exists || depth >= factory.MaxRecursionDepth) { yield break; }
+            foreach (var fileSystemInfo in this.factory.SafeEnumerateFileSystemInfos(this.concreteDirectoryInfo))
+            {
+                var directoryInfo = fileSystemInfo as DirectoryInfo;
+                if (directoryInfo != null)
+                {
+                    yield return this.factory.CreateDirectoryInfoWrapper(directoryInfo, depth + 1);
+                }
+                else
+                {
+                    yield return this.factory.CreateFileInfoWrapper((FileInfo)fileSystemInfo, depth);
+                }
```

Comment:
This could all probably be a switch using pattern matching:

```csharp
switch (fileSystemInfo) {
    case DirectoryInfo directoryInfo:
        yield return this.factory.CreateDirectoryInfoWrapper(directoryInfo, depth + 1);
    case FileInfo fileSystemInfo:
        yield return this.factory.CreateFileInfoWrapper((FileInfo)fileSystemInfo, depth);
    default:
        // this is unreachable, right?
        break;
}
```

Reply:
Code snippet doesn't work :-(  Also C# doesn't seem to have inbuilt switch-on-type

Reply:
It's a thing. See the [second example here](https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching#using-pattern-matching-switch-statements).

I probably mistyped it.

Reply:
Might not work because the second var is `fileSystemInfo` which is already declared.  Try changing to `fileInfo`.  Also might need a normal `break` after the `yield return`.

If not, you can still use and `is` assignment to get a little cleaner:

```csharp
if (fileSystemInfo is DirectoryInfo directoryInfo)
{
    yield return this.factory.CreateDirectoryInfoWrapper(directoryInfo, depth + 1);
    continue;
}

yield return _factory.CreateFileInfoWrapper((FileInfo)fileSystemInfo, _depth);
```

Reply:
Done